### PR TITLE
Better Mod/Plugin dependencies handling

### PR DESCRIPTION
**Pull request by Phinner**
Currently, all plugins are loaded, instantiated and then sorted, this leads to `ClassNotFound` exceptions if you dare to use static classes of other Plugins outside the `init` or if you try to create a general purpose `Plugin` classes like a `AbstractGamemodePlugin` for example that may be optimized for game mode plugins.

The solution is simple, resolving dependencies before loading anything, it features:
- more complex dependency trees
- detecting missing dependencies
- preventing circular dependencies

The code is tested and it works fine, It's just that the structure of `Mods` is a bit hard to follow, so I need your opinion on how the missing/circular dependencies should be handled and what to do with the `dependencies` and `missingDependencies` fields of `LoadedMod`.

Thanks for reading.

If your pull request is **not** translation or serverlist-related, read the list of requirements below and check each box:

- [x] I have read the [contribution guidelines](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md).
- [x] I have ensured that my code compiles, if applicable.
- [x] I have ensured that any new features in this PR function correctly in-game, if applicable.

### CODE REVIEW

**Comment by sk7725** [comment]
try the new loading structure with other mods, especially mods with UI / java mods with static references all over the place / asset-heavy java or js mods that rely on asset loading (and related events) order.
Mods like Sharustry, BetaMindy, and NH should be tested.

(for one, I know BM is heavily dependant on the order of ClientLoad, static init, mod init(), mod loadContent(), and the first Trigger.update, etc. for most of the music / sound filter / UI overlay handling, and they won't get caught by Anuke's BM check because the run is headless)

**Comment by Phinner** [comment]
> try the new loading structure with other mods, especially mods with UI / java mods with static references all over the place / asset-heavy java or js mods that rely on asset loading (and related events) order.
> Mods like Sharustry, BetaMindy, and NH should be tested.
> 
> (for one, I know BM is heavily dependant on the order of ClientLoad, static init, mod init(), mod loadContent(), and the first Trigger.update, etc. for most of the music / sound filter / UI overlay handling, and they won't get caught by Anuke's BM check because the run is headless)

I tried with fakes mods to see if it orders correctly and others that I already have such as `ui-lib` or `BrainDustry`. I will try the one you mentioned later, thanks for pointing it out.

**Comment by sk7725** [comment]
(try BM's latest action, the release is far too outdated)

**Comment by Phinner** [comment]
@sk7725 , I achieved crash-less-dustry :^)

image

**Comment by Phinner** [comment]
@Anuken Ready for review

**Comment by Phinner** [comment]
;-; No review ? ;-;

**Comment by Anuken** [comment]
I'm fairly busy with the v7 branch at the moment.

**Comment by Phinner** [comment]
ok

File: core/src/mindustry/mod/Mods.java [hunk 1]
```diff
@@ -708,6 +711,82 @@ public void contextRun(LoadedMod mod, Runnable run){
...
+
+        ModMeta meta = json.fromJson(ModMeta.class, Jval.read(metaFile.readString()).toString(Jformat.plain));
+        meta.cleanup();
+        return meta;
+    }
+
+    /** Resolves the loading order of a list mods/plugins using their internal names.
+     * It also skips non-mods files or folders. */
+    public Seq<Fi> resolveDependencies(Seq<Fi> files){
+        ObjectMap<String, Fi> fileMapping = new ObjectMap<>(files.size);
```
Comment (Anuken) [inline core/src/mindustry/mod/Mods.java:735]: Do not add the `files.size` argument to map/set constructors; `ObjectMap` capacity is significantly more complex and you're just making it slower here.

File: core/src/mindustry/mod/Mods.java [hunk 2]
```diff
@@ -708,6 +711,82 @@ public void contextRun(LoadedMod mod, Runnable run){
...
+        ObjectMap<String, Seq<String>> dependencies = new ObjectMap<>(files.size);
+
+        for(Fi file : files){
+            Fi zip = file.isDirectory() ? file : new ZipFi(file);
+
+            if(zip.list().length == 1 && zip.list()[0].isDirectory()){
+                zip = zip.list()[0];
+            }
+
+            ModMeta meta = findMeta(zip);
```
Comment (Anuken) [inline core/src/mindustry/mod/Mods.java:745]: findMeta can throw an exception due to malformed JSON. You should try/catch here and continue if it throws.

**Review by Anuken — CHANGES_REQUESTED** [review]

**Comment by Anuken** [comment]
Looks good, will test the changes with some mods later.

**Review by Anuken — APPROVED** [review]

**Comment by Anuken** [comment]
image

*why is this a button*
